### ISSUE TITLE
Jumping to different places when scrolling horizontally using touchpad

### ISSUE THREAD

**Issue by ockamey** [NationalSecurityAgency/ghidra#7024]
I noticed that when I'm scrolling horizontally using touchpad (swiping two fingers to the left or right), the screen is jumping so quickly as you can see on the video. Do you know if it's a feature or a bug? Or how can I disable it? I'm using Ghidra 11.1.2 on Ubuntu 24.04 with Gnome.

https://github.com/user-attachments/assets/9536cc12-3168-4738-ba2d-ef45c3dedf2b

**Comment by ryanmkurtz** [comment]
Can you try switching themes to `Nimbus` to see if it changes the behavior?

**Comment by ockamey** [comment]
Thank you for the response!
Unfortunately it's exactly the same. I've attached the video (of a slightly smaller executable).

https://github.com/user-attachments/assets/3af6c5c9-fc76-462a-ad9a-f9a585633d6f

**Comment by octagons** [comment]
I am also affected by this issue. It seems like the mouse scroll senstivity is too high and moving your fingers slightly left or right as you might naturally do when scrolling up or down triggers a forward/backward navigation to the next code segment. If it matters, I'm using a Apple Magic Trackpad over Bluetooth. 

Ghidra 11.1.2 PUBLIC build
OpenJDK 22.0.2
Manjaro Linux, KDE Plasma, Wayland

**Comment by dragonmacher** [comment]
There is a tool option to disable horizontal mouse scrolling.  It can be found under the Code Browser's Edit -> Tool Options, under Options -> Listing Fields -> Mouse -> Horizontal Scrolling.

I am curious to see if turning this off has any effect on this particular issue.

**Comment by octagons** [comment]
@dragonmacher - Just disabled that option and tested but it had no visible effect on the behavior. Notably, horizontal scrolling also doesn't "work" in the sense that it does not scroll the view of a frame whose contents exceed the width of the frame's window whether or not this option is enabled - but it's not clear if this is related I don't want to muddy the issue.

Looking at it more closely, the horizontal scrolling action on the touchpad causes the focus to jump between the navigation history - the same functionality as hitting `alt + left` or `alt + right` on the keyboard.

My system has gesture recognition (similar to a tablet); e.g. a 2-finger swipe in many applications also navigates backwards, a 3-finger swipe moves between virtual desktops, etc. I looked for a few minutes, but cannot find a way to disable the gesture recognition to test whether the behavior persists. If it would help, I can try to test it later when I have more time. @ockamey - Do you also have this gesture recognition feature for your touchpad?

**Comment by dragonmacher** [comment]
@octagons Thanks for the update.    Based on this and other gesture input devices I've seen, I do believe the side scrolling is being interpreted as navigation action presses.

For a normal mouse, users should be able to `Shift-Mouse Wheel` to perform horizontal scrolling.   Does that work for you if you hold shift while performing the vertical scroll gesture?

**Comment by octagons** [comment]
> For a normal mouse, users should be able to `Shift-Mouse Wheel` to perform horizontal scrolling. Does that work for you if you hold shift while performing the vertical scroll gesture?

Yes, I can confirm that the `Shift + mouse wheel` input performs the horizontal scrolling of the frame's contents as you described.

**Comment by dragonmacher** [comment]
Thanks!   I'm guessing that this may have been 'broken' when we added support for mouse button navigation by assigning mouse button bindings to the history navigation actions.   I think you both can fix this issue by clearing the mouse bindings in the options for the actions in the image below:"

![image](https://github.com/user-attachments/assets/755a3dd3-1d34-465c-802d-3903b12375bb)

If you decided to try this, let us know if it works.

**Comment by ockamey** [comment]
@dragonmacher thank you for the response! Unfortunately, the checkbox "Enter Mouse Binding" by default was unchecked on my machine in both previous, and next actions.

**Comment by dragonmacher** [comment]
It sounds like there may be multiple issues happening.  

@octagons Once the mouse bindings were cleared, did you still see the navigation history actions firing?  I'm assuming that issue went away, but there may an issue with processing mouse scrolling events at high scale. 

> but are you aware of any debugging functionality in Ghidra or for Wayland that might show the inputs being sent from the touchpad?   

This exists in the codebase, but is not something you can use from the UI.  If you are setup to make and test changes to our source, then I can show you where you can make a change to get this type of info.

**Comment by dragonmacher** [comment]
> @dragonmacher thank you for the response! Unfortunately, the checkbox "Enter Mouse Binding" by default was unchecked on my machine in both previous, and next actions.

@ockamey The checkbox is not selected by default.  The actions do have a keybinding applied, even though the box is not checked.  To clear the binding, you would first check the box and then clear the keybinding.  @octagons tried this and said it did not help, so this may not fix this issue for you.

**Comment by octagons** [comment]
> This exists in the codebase, but is not something you can use from the UI. If you are setup to make and test changes to our source, then I can show you where you can make a change to get this type of info.

Sure, I'll setup a dev environment tonight/tomorrow. I should be able to replicate my work environment pretty closely on my personal machine.

**Comment by dragonmacher** [comment]
If you end up doing this, you can change this class:

```java
class DockingWindowsContextSensitiveHelpListener {

    private static DockingWindowsContextSensitiveHelpListener instance;
    
    private DockingWindowsContextSensitiveHelpListener() {
        Toolkit toolkit = Toolkit.getDefaultToolkit();
        AWTEventListener listener = new AWTEventListener() {            
            public void eventDispatched( AWTEvent event ) {
                DockingWindowManager.setMouseOverObject( event.getSource() );
            }
        }; 
        toolkit.addAWTEventListener( listener, AWTEvent.MOUSE_MOTION_EVENT_MASK );
    }

    static synchronized void install() {
        if ( instance == null ) {
            instance = new DockingWindowsContextSensitiveHelpListener();
        }
    }
}

```

You should be able to add a line to do this:

```java
toolkit.addAWTEventListener( listener, AWTEvent.MOUSE_MOTION_WHEEL_MASK );
```

This is a class we use to listen to all mouse movement events.  After adding this code to be notified of mouse wheel events, you should be able to print out each event to see what Java is giving to you as you are using your various gestures.

**Comment by ockamey** [comment]
> > @dragonmacher thank you for the response! Unfortunately, the checkbox "Enter Mouse Binding" by default was unchecked on my machine in both previous, and next actions.
> 
> @ockamey The checkbox is not selected by default. The actions do have a keybinding applied, even though the box is not checked. To clear the binding, you would first check the box and then clear the keybinding. @octagons tried this and said it did not help, so this may not fix this issue for you.

You're right, it didn't help. But also I noticed that if I check one of the checkboxes, the another one is automatically checked as well. Also after checking them, I closed the exit window and opened it again, and then the checkboxes are still not checked.

**Comment by Tropix126** [comment]
This seems like the same root-cause of the bug that JADX had in https://github.com/skylot/jadx/issues/1718#issuecomment-1660648574

**Comment by cyrozap** [comment]
I'm not the OP, but I can confirm that following @d235j's instructions fixed this issue for me. I was experiencing this issue when scrolling in various windows using the TrackPoint nub on my ThinkPad. I'm running Ghidra 11.3.1 on Arch Linux in a KDE Plasma Wayland session.

**Comment by PennRobotics** [comment]
Restarting Ghidra was the necessary step for me. If you're stuck and haven't restarted&hellip;

**Comment by d235j** [comment]
@dragonmacher is any further feedback needed here? I think I managed to find the root cause, see above. I suggest just removing those mouse bindings from the defaults entirely.

**Comment by dragonmacher** [comment]
@d235j I think what you posted above is a good workaround.  It may not work for everyone though.  Either way, I don't see anything we can do to fix this on our side.

**Comment by d235j** [comment]
I think part of the issue is that the checkbox to change the mouse bindings is very confusing. Maybe there should be a second field for the mouse bindings, instead of the checkbox.

**Comment by dragonmacher** [comment]
@d235j I think you are right that the UI could be improved.  I will look into that.
